PERF: Faster SparseArray.__getitem__ for boolean masks (#23122)

bdrum opened the pull request:
- [ ] closes #23122
- [ ] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [ ] The issue #23122 contains recommendation from @TomAugspurger to avoid densifying when boolean sparse array is used as index. I have tried few solutions including usage direct sp_index.indices in a case of fill_value is False and np.setdiff1d for opposite case.
Usage of direct indexing is working very fast, but calculating setdiff1d is working very slow, the good solution for both case (much faster than setdiff1d and bit slower than direct indices) using masks. Here is the plot of comparison old (densifying) implementation and new one (masks):
![](https://sun9-21.userapi.com/impg/sQrZSIohCDr3EesWoWhNALJHeBZxe7JdOmkISA/kvYTPuRRfeY.jpg?size=2560x1580&quality=96&sign=b7e176a5bd97154b47ebd30b01de3035&type=album)
I understand that using mask will take more memory, but I don't think that it is crucial point here especially for bool8 type, but let's discuss.
We can combine both approaches and use direct indices when fill value is False and use mask in the opposite case.
One more question to discuss related with case when True value of Boolean sparse array lead to the fill value of original array. I assume that in this case fill value should be return.

I've got benchmark results via pytest-benchmark framework.

Code for benchmarking:

~~~python
import pytest
from pandas.arrays import SparseArray
import numpy as np


class TestPerf:
    def setup_method(self, method):
        d = 0.8
        self.n = 1_000_000
        self.data = np.random.rand(self.n)
        genInd = lambda d: np.unique(
            np.random.randint(
                low=0, high=self.n - 1, size=int(self.n * d), dtype=np.int32
            )
        )
        self.inds = genInd(d)
        barrs = np.full(shape=self.n, fill_value=True, dtype=np.bool8)
        arr_dens_inds = genInd(d)

        self.data[self.inds] = np.nan
        barrs[arr_dens_inds] = False

        self.sp_arrs = SparseArray(self.data)
        self.sp_barrs_false = SparseArray(barrs, dtype=np.bool8, fill_value=False)
        self.sp_barrs_true = SparseArray(~barrs[d], dtype=np.bool8, fill_value=True)

    def perf_sparse_boolean_indexing_fv_true(self):
        return self.sp_arrs[self.sp_barrs_true]

    def perf_sparse_boolean_indexing_fv_false(self):
        return self.sp_arrs[self.sp_barrs_false]

    def test_perf_sparse_boolean_indexing_fv_true(self, benchmark):
        benchmark(self.perf_sparse_boolean_indexing_fv_true)

    def test_perf_sparse_boolean_indexing_fv_false(self, benchmark):
        benchmark(self.perf_sparse_boolean_indexing_fv_false)

~~~

bdrum committed:
PERF: Faster SparseArray.__getitem__ for boolean masks (#23122)

bdrum commented on pandas/core/arrays/sparse/array.py:
```diff
@@ -945,13 +949,18 @@ def __getitem__(
             if isinstance(key, SparseArray):
                 if is_bool_dtype(key):
-                    key = key.to_dense()
+                    if is_bool(key.fill_value):
```
Here I'm not sure that such check is not redundant.  According to logic SparseDType(bool) should have bool fill_value, but in fact it could be an any value..

jbrockmendel replied:
there's an issue somewhere about being stricter about what fill_values are allowed for a given dtype, would be great to see an improvement there

bdrum replied:
Hm..., actually right now I've added a changes to SparseDtype that will rise an exception in case of non boolean fill_value for SparseDtype(bool), but then I think what if user have an array with bool values and nan as fill_value? In this case we have to keep current functionality.

jbrockmendel commented on pandas/core/arrays/sparse/array.py:
```diff
@@ -701,7 +701,11 @@ def isna(self):
         dtype = SparseDtype(bool, self._null_fill_value)
-        return type(self)._simple_new(isna(self.sp_values), self.sp_index, dtype)
+        if self._null_fill_value:
+            return type(self)._simple_new(isna(self.sp_values), self.sp_index, dtype)
```
do we still need to be concerned about nulls in the sp_values?

bdrum replied:
I agreed that is looks redundant, but now I can reproduce the case when without this check the result will be wrong.
It related with [the issue](https://github.com/pandas-dev/pandas/issues/44956).
~~~python
s = pd.arrays.SparseArray([np.nan, 1,2,3,0,0,0, np.nan], fill_value=0)
#[nan, 1.0, 2.0, 3.0, 0, 0, 0, nan]
#Fill: 0
#IntIndex
#Indices: array([0, 1, 2, 3, 7])
s.fill_value = np.nan
s
#[nan, 1.0, 2.0, 3.0, nan, nan, nan, nan]
#Fill: nan
#IntIndex
#Indices: array([0, 1, 2, 3, 7])

pd.isna(s)
#[True, False, False, False, True, True, True, True]
#Fill: True
#IntIndex
#Indices: array([0, 1, 2, 3, 7])
~~~

By the way, for this case indices of last array are worng. To be fixed and tested...

jbrockmendel replied:
> s.fill_value = np.nan

I didn't realize this was settable.  Should definitely deprecate that.

I guess without `__setitem__` its tough to get cases where you have both a nan fill_value and nan in the sp_values?

bdrum replied:
Setting of fill_value doesn't use __setitem__. E.g. for mentioned case we can't use __setitem__:

~~~python
 s[0] = 0
---------------------------------------------------------------------------
TypeError: SparseArray does not support item assignment via setitem
~~~

I'm not sure is this correct to deprecate changing fill_value? 
Perhaps we have to add some value checking depending from type and nan...

jreback commented on pandas/tests/extension/test_sparse.py:
```diff
@@ -208,6 +208,12 @@ def test_isna(self, data_missing):
+        data_missing = data_missing.fillna(0)
+        expected_dtype = SparseDtype(bool, pd.isna(data_missing.dtype.fill_value))
+        expected = SparseArray([False, False], fill_value=False, dtype=expected_dtype)
+        self.assert_equal(pd.isna(data_missing), expected)
```
use .isna() here (the method not the function)

jreback replied:
use result= for this

jreback commented on pandas/tests/arrays/sparse/test_array.py:
```diff
@@ -248,6 +248,16 @@ def test_scalar_with_index_infer_dtype(self, scalar, dtype):
+
+        spar_bool = SparseArray(~spar_bool.to_dense(), dtype=np.bool8, fill_value=False)
+        exp = SparseArray([np.nan, 1, 3, 4, np.nan])
+        tm.assert_sp_array_equal(self.arr[spar_bool], exp)
```
use res= for the result and leave the comparison with just res, exp

jreback reviewed (changes requested):
can you add a whatsnew note for the perf improvement. is there a bug fix here too?
do we have asv's that cover these cases?

jreback commented:
@bdrum fine on the bugfix, just add a whatsnew note describing (and you can reference this PR number)

bdrum commented:
> @bdrum fine on the bugfix, just add a whatsnew note describing (and you can reference this PR number)

@jreback Thanks! 
Also I've made a changes again according to discussion with @jbrockmendel 
And fixed bugs with indexes after applying unary operations.

What about increasing performance. I've added benchmark to asv, but looks strange
~~~
       before           after         ratio
     [47eb2198]       [1bb10c75]
     <master>         <dev23122>
+      22.3±0.5ms         44.0±4ms     1.97  sparse.GetItemMask.time_mask(True)
-     1.62±0.05ms         48.9±8μs     0.03  sparse.GetItemMask.time_mask(False)
~~~

Because when I ran the same code via timeit, I've got huge performance increasing for False fill_value and a bit (about 5-7%) when fill_value is True.

bdrum commented:
Sorry, a bit confusion after merging conflicts...

Now I have 26 failed test in my branch as well as in current master:

<details>
=============================================================== short test summary info ===============================================================
FAILED pandas/tests/arithmetic/test_datetime64.py::TestDatetime64ArrayLikeComparisons::test_dt64arr_cmp_arraylike_invalid[numexpr-tzlocal()-Index-other3]
FAILED pandas/tests/arithmetic/test_datetime64.py::TestDatetime64ArrayLikeComparisons::test_dt64arr_cmp_arraylike_invalid[numexpr-tzlocal()-Index-other7]
FAILED pandas/tests/arithmetic/test_datetime64.py::TestDatetime64ArrayLikeComparisons::test_dt64arr_cmp_arraylike_invalid[numexpr-tzlocal()-Index-other9]
FAILED pandas/tests/arithmetic/test_datetime64.py::TestDatetime64ArrayLikeComparisons::test_dt64arr_cmp_arraylike_invalid[numexpr-tzlocal()-Series-other3]
FAILED pandas/tests/arithmetic/test_datetime64.py::TestDatetime64ArrayLikeComparisons::test_dt64arr_cmp_arraylike_invalid[numexpr-tzlocal()-Series-other6]
FAILED pandas/tests/arithmetic/test_datetime64.py::TestDatetime64ArrayLikeComparisons::test_dt64arr_cmp_arraylike_invalid[numexpr-tzlocal()-Series-other7]
FAILED pandas/tests/arithmetic/test_datetime64.py::TestDatetime64ArrayLikeComparisons::test_dt64arr_cmp_arraylike_invalid[numexpr-tzlocal()-Series-other9]
FAILED pandas/tests/arithmetic/test_datetime64.py::TestDatetime64ArrayLikeComparisons::test_dt64arr_cmp_arraylike_invalid[numexpr-tzlocal()-DataFrame-other3]
FAILED pandas/tests/arithmetic/test_datetime64.py::TestDatetime64ArrayLikeComparisons::test_dt64arr_cmp_arraylike_invalid[numexpr-tzlocal()-array-other3]
FAILED pandas/tests/arithmetic/test_datetime64.py::TestDatetime64ArrayLikeComparisons::test_dt64arr_cmp_arraylike_invalid[numexpr-tzlocal()-array-other7]
FAILED pandas/tests/arithmetic/test_datetime64.py::TestDatetime64ArrayLikeComparisons::test_dt64arr_cmp_arraylike_invalid[numexpr-tzlocal()-array-other9]
FAILED pandas/tests/arithmetic/test_datetime64.py::TestDatetime64ArrayLikeComparisons::test_dt64arr_cmp_mixed_invalid[numexpr-tzlocal()] - OSError: [...
FAILED pandas/tests/arithmetic/test_datetime64.py::TestDatetime64ArrayLikeComparisons::test_dt64arr_cmp_arraylike_invalid[python-tzlocal()-Index-other3]
FAILED pandas/tests/arithmetic/test_datetime64.py::TestDatetime64ArrayLikeComparisons::test_dt64arr_cmp_arraylike_invalid[python-tzlocal()-Index-other7]
FAILED pandas/tests/arithmetic/test_datetime64.py::TestDatetime64ArrayLikeComparisons::test_dt64arr_cmp_arraylike_invalid[python-tzlocal()-Index-other9]
FAILED pandas/tests/arithmetic/test_datetime64.py::TestDatetime64ArrayLikeComparisons::test_dt64arr_cmp_arraylike_invalid[python-tzlocal()-Series-other3]
FAILED pandas/tests/arithmetic/test_datetime64.py::TestDatetime64ArrayLikeComparisons::test_dt64arr_cmp_arraylike_invalid[python-tzlocal()-Series-other6]
FAILED pandas/tests/arithmetic/test_datetime64.py::TestDatetime64ArrayLikeComparisons::test_dt64arr_cmp_arraylike_invalid[python-tzlocal()-Series-other7]
FAILED pandas/tests/arithmetic/test_datetime64.py::TestDatetime64ArrayLikeComparisons::test_dt64arr_cmp_arraylike_invalid[python-tzlocal()-Series-other9]
FAILED pandas/tests/arithmetic/test_datetime64.py::TestDatetime64ArrayLikeComparisons::test_dt64arr_cmp_arraylike_invalid[python-tzlocal()-DataFrame-other3]
FAILED pandas/tests/arithmetic/test_datetime64.py::TestDatetime64ArrayLikeComparisons::test_dt64arr_cmp_arraylike_invalid[python-tzlocal()-array-other3]
FAILED pandas/tests/arithmetic/test_datetime64.py::TestDatetime64ArrayLikeComparisons::test_dt64arr_cmp_arraylike_invalid[python-tzlocal()-array-other7]
FAILED pandas/tests/arithmetic/test_datetime64.py::TestDatetime64ArrayLikeComparisons::test_dt64arr_cmp_arraylike_invalid[python-tzlocal()-array-other9]
FAILED pandas/tests/arithmetic/test_datetime64.py::TestDatetime64ArrayLikeComparisons::test_dt64arr_cmp_mixed_invalid[python-tzlocal()] - OSError: [E...
FAILED pandas/tests/indexing/test_chaining_and_caching.py::TestChaining::test_detect_chained_assignment_warning_stacklevel[3] - AssertionError: asser...
FAILED pandas/tests/indexing/test_chaining_and_caching.py::TestChaining::test_detect_chained_assignment_warning_stacklevel[rhs1] - AssertionError: as...
====================================== 26 failed, 4 passed, 2 skipped, 151259 deselected, 29 warnings in 52.32s =======================================
<details>

jbrockmendel commented:
Can you post the traceback for one of the arithmetic tests that failed?

bdrum commented:
> Can you post the traceback for one of the arithmetic tests that failed?

Sure. Here it is:

<details>
<?xml version="1.0" encoding="utf-8"?><testsuites><testsuite name="pytest" errors="0" failures="1" skipped="0" tests="1" time="0.974" timestamp="2021-12-22T10:48:13.621099" hostname="RUMLAB"><testcase classname="pandas.tests.arithmetic.test_datetime64.TestDatetime64ArrayLikeComparisons" name="test_dt64arr_cmp_arraylike_invalid[numexpr-tzlocal()-Index-other3]" time="0.012"><failure message="OSError: [Errno 22] Invalid argument">self = &lt;pandas.tests.arithmetic.test_datetime64.TestDatetime64ArrayLikeComparisons object at 0x000001F5FE6B3790&gt;
other = array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9], dtype=object)
tz_naive_fixture = tzlocal()
box_with_array = &lt;class 'pandas.core.indexes.base.Index'&gt;

    @pytest.mark.parametrize(
        "other",
        [
            # GH#4968 invalid date/int comparisons
            list(range(10)),
            np.arange(10),
            np.arange(10).astype(np.float32),
            np.arange(10).astype(object),
            pd.timedelta_range("1ns", periods=10).array,
            np.array(pd.timedelta_range("1ns", periods=10)),
            list(pd.timedelta_range("1ns", periods=10)),
            pd.timedelta_range("1 Day", periods=10).astype(object),
            pd.period_range("1971-01-01", freq="D", periods=10).array,
            pd.period_range("1971-01-01", freq="D", periods=10).astype(object),
        ],
    )
    def test_dt64arr_cmp_arraylike_invalid(
        self, other, tz_naive_fixture, box_with_array
    ):
        tz = tz_naive_fixture
    
        dta = date_range("1970-01-01", freq="ns", periods=10, tz=tz)._data
        obj = tm.box_expected(dta, box_with_array)
&gt;       assert_invalid_comparison(obj, other, box_with_array)

pandas\tests\arithmetic\test_datetime64.py:122: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pandas\tests\arithmetic\common.py:111: in assert_invalid_comparison
    result = xbox2(left == right)
pandas\core\ops\common.py:70: in new_method
    return method(self, other)
pandas\core\arraylike.py:40: in __eq__
    return self._cmp_method(other, operator.eq)
pandas\core\indexes\base.py:6591: in _cmp_method
    result = op(self._values, other)
pandas\core\ops\common.py:70: in new_method
    return method(self, other)
pandas\core\arraylike.py:40: in __eq__
    return self._cmp_method(other, operator.eq)
pandas\core\arrays\datetimelike.py:1024: in _cmp_method
    op, np.asarray(self.astype(object)), other
pandas\core\arrays\datetimes.py:666: in astype
    return dtl.DatetimeLikeArrayMixin.astype(self, dtype, copy)
pandas\core\arrays\datetimelike.py:413: in astype
    converted = ints_to_pydatetime(
pandas\_libs\tslibs\vectorized.pyx:177: in pandas._libs.tslibs.vectorized.ints_to_pydatetime
    local_value = tz_convert_utc_to_tzlocal(value, tz)
pandas\_libs\tslibs\tzconversion.pyx:383: in pandas._libs.tslibs.tzconversion.tz_convert_utc_to_tzlocal
    return _tz_convert_tzlocal_utc(utc_val, tz, to_utc=False, fold=fold)
pandas\_libs\tslibs\tzconversion.pyx:594: in pandas._libs.tslibs.tzconversion._tz_convert_tzlocal_utc
    delta = _tzlocal_get_offset_components(val, tz, to_utc, fold)
pandas\_libs\tslibs\tzconversion.pyx:554: in pandas._libs.tslibs.tzconversion._tzlocal_get_offset_components
    dt = dt.astimezone(tz)
..\..\..\Anaconda3\envs\pandas-dev\lib\site-packages\dateutil\tz\_common.py:144: in fromutc
    return f(self, dt)
..\..\..\Anaconda3\envs\pandas-dev\lib\site-packages\dateutil\tz\_common.py:261: in fromutc
    _fold = self._fold_status(dt, dt_wall)
..\..\..\Anaconda3\envs\pandas-dev\lib\site-packages\dateutil\tz\_common.py:196: in _fold_status
    if self.is_ambiguous(dt_wall):
..\..\..\Anaconda3\envs\pandas-dev\lib\site-packages\dateutil\tz\tz.py:254: in is_ambiguous
    naive_dst = self._naive_is_dst(dt)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = tzlocal(), dt = datetime.datetime(1969, 12, 31, 21, 0, tzinfo=tzlocal())

    def _naive_is_dst(self, dt):
        timestamp = _datetime_to_timestamp(dt)
&gt;       return time.localtime(timestamp + time.timezone).tm_isdst
E       OSError: [Errno 22] Invalid argument

..\..\..\Anaconda3\envs\pandas-dev\lib\site-packages\dateutil\tz\tz.py:260: OSError</failure></testcase></testsuite></testsuites>

<details>

bdrum commented:
asv report:

~~~
       before           after         ratio
     [2cc9ab3e]       [4f58c8a8]
     <master>         <dev23122>
-     1.69±0.01ms         46.4±1μs     0.03  sparse.GetItemMask.time_mask(False)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
~~~

<details>

~~~
[ 50.00%] · For pandas commit 4f58c8a8 <dev23122> (round 2/2):
[ 50.00%] ·· Benchmarking conda-py3.8-Cython0.29.24-jinja2-matplotlib-numba-numexpr-numpy-odfpy-openpyxl-pyarrow-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[ 75.00%] ··· sparse.GetItemMask.time_mask                                                                                                           ok
[ 75.00%] ··· ============ ============
               fill_value
              ------------ ------------
                  True      24.3±0.2ms
                 False       46.4±1μs
                  nan       22.1±0.5ms
              ============ ============

[ 75.00%] · For pandas commit 2cc9ab3e <master> (round 2/2):
[ 75.00%] ·· Benchmarking conda-py3.8-Cython0.29.24-jinja2-matplotlib-numba-numexpr-numpy-odfpy-openpyxl-pyarrow-pytables-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[100.00%] ··· sparse.GetItemMask.time_mask                                                                                                           ok
               fill_value
              ------------ -------------
                  True       24.8±0.8ms
                 False      1.69±0.01ms
                  nan        23.6±0.5ms
              ============ =============
~~~

jreback commented:
@bdrum how's this coming?

bdrum commented:
@jreback Thanks!) 

I think that is done right now. I resolved your changes request: modified whatsnew file and add benchmark to asv.
So do you have any comments or remarks?

jreback commented:
some failing builds which are relevant benchmarks and i think the doc build

bdrum commented:
Ah, you right that's my fault. Doc doesn't know what is pull, in this case as I can see I have to use :issue: and number of this pr. Sorry about that.

I just noticed that some pr was merged despite on the fact that not all check was done, so that is the reason why I wasn't so attentive about that. Now I will. 

Let me fix it.

jreback commented:
> I just noticed that some pr was merged despite on the fact that not all check was done, so that is the reason why I wasn't very attentive about that. Now I will.

well i check the failures to see if they are unrelated but generally we need all green to avoid issues

bdrum commented:
@jreback Now failed checks looks like not by my fault.

jreback commented:
thanks @bdrum

phofl commented:
This broke some things with greater comparisons. We should fix them before 1.4 is released or revert here.

```
s = pd.arrays.SparseArray([1, 2, 3, 4, np.nan, np.nan], fill_value=np.nan)
s[s>2]
```
returned

```
[3.0, 4.0]
Fill: nan
IntIndex
Indices: array([0, 1], dtype=int32)
```

which was correct. Now this returns 

```
[1.0, 2.0, 3.0, 4.0]
Fill: nan
IntIndex
Indices: array([0, 1, 2, 3], dtype=int32)
```
